### PR TITLE
Add origin to ip addresses

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

Adding an origin field to all ip address with upgrade steps to always be
provider by default, should allow us to reason about whom owns the ip
address.

Once we know about the owner of the ip address we can then think about
the deletion of the ip address if it's no longer used.

The upgrade steps in this PR ensure that we always have on and that it's
always set to the default origin; Provider.

----

Technically we expect the machiner (worker) to never delete any ip
addresses it doesn't own (origin in this case). So if a machiner sees
that the owner is a "provider" we skip this. If it sees the owner is a
"machine" and the ip address is no longer in use, the machiner can then
delete it from the database.

## QA steps

 - Bootstrap 2.7 and deploy a workload.
 - Upgrade to 2.8 and ensure that we now have the default `origin` on ip addresses.  

## Bug reference

This doesn't solve the bug but starts to land the foundations for fixing it.

https://bugs.launchpad.net/juju/+bug/1733968

### CODE REVIEW

**Comment by SimonRichardson** [comment]
https://trello.com/c/B4aYxUre/2487-add-origin-to-ip-addresses

File: state/linklayerdevices_ipaddresses.go [hunk 1]
```diff
@@ -63,6 +63,15 @@ type ipAddressDoc struct {
 
 	// IsDefaultGateway is set to true if that device/subnet is the default gw for the machine
 	IsDefaultGateway bool `bson:"is-default-gateway,omitempty"`
+
+	// Origin represents whom owns or identifies with this ipAddress.
```
Comment (manadart) [inline state/linklayerdevices_ipaddresses.go:67]: "represents the authoritative source of..."

File: state/linklayerdevices_ipaddresses.go [hunk 2]
```diff
@@ -196,11 +205,21 @@ func (addr *Address) GatewayAddress() string {
...
 }
 
-// IsDefaultGateway returns true if this address is used for the default gw on the machine.
+// IsDefaultGateway returns true if this address is used for the default gw on
+// the machine.
 func (addr *Address) IsDefaultGateway() bool {
 	return addr.doc.IsDefaultGateway
 }
 
+// Origin represents whom owns or identifies with this ipAddress.
```
Comment (manadart) [inline state/linklayerdevices_ipaddresses.go:214]: Same here.

File: state/upgrades.go [hunk 3]
```diff
@@ -2923,6 +2923,49 @@ func AddMachineIDToSubordinates(pool *StatePool) error {
...
+		// Potentially we should check if it's a valid Origin and if it's not
+		// set it to the default origin.
+		if ipAddress.Origin != "" {
+			continue
+		}
+
+		ops = append(ops, txn.Op{
+			C:      ipAddressesC,
+			Id:     ipAddress.DocID,
+			Update: bson.M{"$set": bson.M{"origin": network.OriginProvider}},
```
Comment (manadart) [inline state/upgrades.go:2959]: Perhaps a note as to why we set it to provider for safety and later quiescence?

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**